birdvoxpaint/util.py

```diff
@@ -0,0 +1,225 @@
+
+    # only feed in full length frames into the pipeline
+    truncated_length = (S.shape[1] // segment_length) * segment_length
+    S = S[:,:truncated_length] # removes incomplete frames
+    S = S.T.reshape(-1, segment_length, len(S)) # shape: frame, segment, freq
+    S = S.transpose((0, 2, 1)) # shape: frame, freq, segment
+    return S
+
+
+
+def apply_indices(S, indices, segment_length, n_jobs=1):
+    '''Takes an iterable yielding blocks of spectrograms, breaks them into blocks of
+    size: segment_length. The number of blocks per yield should equal n_jobs.
+
+    Arguments:
+        S (iterable(np.ndarray[freq, time])): the iterable of spectrograms
+        indices (callable/list of callables): the indices to calculate. If a
+            single callable, it should return all indices as the final axis.
+        segment_length (int): the length in time (# frames) to be passed to
+            indices at a time.
+
+    '''
+    # reshape to blocks of spectrograms (# blocks == n_jobs)
+    S_blocks = (spec_frame(spec, segment_length) for spec in S) # yields: block, freq, time
+
+    # allow user to pass a custom indices function - or as a list of functions
+    if callable(indices):
+        calc_indices = indices
+    else:
+        calc_indices = lambda x: np.stack([f(x) for f in indices], axis=-1)
+
+    # build job to calculate indices
+    if n_jobs == 1:
+        jobs = joblib.Parallel(n_jobs=n_jobs)
+        calc_indices = joblib.delayed(calc_indices)
+    else:
+        jobs = lambda x: x # dummy function so that we can call jobs below like normal.
+
+    S_blocks = (
+        np.stack(list(jobs(map(calc_indices, spec))), axis=1) # shape: freq, block
+        for spec in S_blocks # yields: block, freq, time
+        if spec.size # removes empty blocks
+    )
+
+    S = np.concatenate(list(S_blocks), axis=1) # shape: freq, block, indices
+    return S
+
+def _shape(X):
+    for x in X:
+        print(x.shape)
+        yield x
+
+def freq_slice(fmin, fmax, sr, n_fft):
+    '''Calculate the slice needed to select a frequency band.
+
+    Arguments:
+        fmin, fmax (int): the frequency bounds
+        sr (int): the sample rate
+        n_fft (int): the fft size
+
+    Returns:
+        slice(i[fmin], i[fmax])
+    '''
+    if not sr or not n_fft:
+        raise ParameterError(
+            "You must set a sr=({}) and n_fft=({})".format(sr, n_fft))
+
+    if fmin and fmin < 0:
+        raise ParameterError("fmin={} must be nonnegative".format(fmin))
+
+    if fmax and fmax > (sr/2):
+        raise ParameterError(
+            "fmax={} must be smaller than sample rate sr={}".format(fmax, sr))
+
+    fft_frequencies = librosa.fft_frequencies(sr=sr, n_fft=n_fft)
+    bin_start = np.where(fft_frequencies >= fmin)[0][0] if fmin else None
+    bin_stop = np.where(fft_frequencies < fmax)[0][-1] if fmax else None
+    return slice(bin_start, bin_stop)
+
+
+
+DAY = 60*60*24
+MONTH = DAY * 30
+def format_based_on_scale(t, scale, offset=None):
```

Comment:
use `librosa.frames_to_time` or the `TimeFormatter` decorator

Reply:
Oh werd. I didn't realize Librosa had a time formatter. It looks like it doesn't support multi day/month long stuff though. I also want to be able to show absolute dates if that's what the time scale is on. Like if we're showing 6 months, I'd want it to show Jan, Feb, ... not `00:00:00 - 719:59:59` (or `23:59:59` or whatever).